plugins/catalog-backend/src/ingestion/processors/AnnotateLocationEntityProcessor.ts

```diff
@@ -15,31 +15,71 @@
  */
 
 import {
+  EDIT_URL_ANNOTATION,
   Entity,
   LocationSpec,
   LOCATION_ANNOTATION,
   ORIGIN_LOCATION_ANNOTATION,
+  SOURCE_LOCATION_ANNOTATION,
   stringifyLocationReference,
+  VIEW_URL_ANNOTATION,
 } from '@backstage/catalog-model';
-import lodash from 'lodash';
+import { ScmIntegrationRegistry } from '@backstage/integration';
+import { identity, merge, pickBy } from 'lodash';
 import { CatalogProcessor, CatalogProcessorEmit } from './types';
 
+type Options = {
+  integrations: ScmIntegrationRegistry;
+};
+
 export class AnnotateLocationEntityProcessor implements CatalogProcessor {
+  constructor(private readonly options: Options) {}
+
   async preProcessEntity(
     entity: Entity,
     location: LocationSpec,
     _: CatalogProcessorEmit,
     originLocation: LocationSpec,
   ): Promise<Entity> {
-    return lodash.merge(
+    const { integrations } = this.options;
+    let viewUrl;
+    let editUrl;
+    let sourceLocation;
+
+    if (location.type === 'url') {
+      const scmIntegration = integrations.byUrl(location.target);
+
+      viewUrl = location.target;
+      editUrl = scmIntegration?.resolveEditUrl(location.target);
+
+      const sourceUrl = scmIntegration?.resolveUrl({
+        url: './',
+        base: location.target,
+      });
+
+      if (sourceUrl) {
+        sourceLocation = stringifyLocationReference({
+          type: 'url',
+          target: sourceUrl,
+        });
+      }
+    }
+
+    return merge(
       {
         metadata: {
-          annotations: {
-            [LOCATION_ANNOTATION]: stringifyLocationReference(location),
-            [ORIGIN_LOCATION_ANNOTATION]: stringifyLocationReference(
-              originLocation,
-            ),
-          },
+          annotations: pickBy(
+            {
+              [LOCATION_ANNOTATION]: stringifyLocationReference(location),
+              [ORIGIN_LOCATION_ANNOTATION]: stringifyLocationReference(
+                originLocation,
+              ),
+              [VIEW_URL_ANNOTATION]: viewUrl,
+              [EDIT_URL_ANNOTATION]: editUrl,
+              [SOURCE_LOCATION_ANNOTATION]: sourceLocation,
+            },
+            identity,
```

Comment:
do you need the pick and identity? isn't the merge a deep merge anyway? 🤔

Reply:
It's to avoid having keys that are filled with `undefined`, as `editUrl`, `viewUrl` and `sourceLocation` might be `undefined`. I think my experiments showed that merge would copy undefined values.

Reply:
Ingestion would that fail with:

```
[1] 2021-03-08T10:59:29.090Z catalog warn Failed item in location file:../../plugins/scaffolder-backend/sample-templates/all-templates.yaml, InputError: Malformed envelope, /metadata/annotations/backstage.io~1view-url should be string
```

Reply:
Ah yes that makes sense! Yeah I have code somewhere else that does like `stripKeysWithUndefinedValues` for the same reason